File: src/dao/ChannelDao.kt

/**
 * 隧道数据操作
 */
object ChannelDao {

    /**
     * 查询所有key
     */
    const val SELECT_KEY = """
     channel.id as id
    ,channel.client_id as clientId
    ,channel.name as name
    ,channel.type as type
    ,channel.server_port as serverPort
    ,channel.target_port as targetPort
    ,channel.in_data_total as inDataTotal
    ,channel.out_data_total as outDataTotal
    ,channel.enable_state as enableState
    ,channel.security_state as securityState
    ,channel.acl_state as aclState
    ,channel.create_date as createDate
    ,channel.update_date as updateDate
    ,channel.remark as remark
    """



    /**
     * 添加一条隧道
     */
    fun add(dto: ChannelDto) {
        val updateDate = System.currentTimeMillis()
        val sql =
            "insert into channel(client_id,name,type,server_port,target_port,security_state,acl_state,enable_state,update_date)values(?,?,?,?,?,?,?,?,?)"
        NPSDB.db.exec(
            sql,
            dto.clientId,
            dto.name,
            dto.type,
            dto.serverPort,
            dto.targetPort,
            dto.securityState,
            dto.aclState,
            dto.enableState,
            updateDate
        )

        val insertIdSql = "select id from channel where update_date = ?"
        val id = NPSDB.db.selectSingleOne<Int>(insertIdSql, updateDate)
        dto.id = id
    }

    /**
     * 通过id获取一条数据
     * @param id 隧道id
     * @return 隧道Dto
     */
    fun selectOne(id: Int): ChannelDto? {
        val sql = "select " + SELECT_KEY + " from channel where id = ?"
        val dto = NPSDB.db.selectOne(ChannelDto::class.java, sql, id)
        return dto
    }

    /**
     * 获取所有数据
     * @return 隧道Dto
     */
    fun selectAll(): List<ChannelDto> {
        val sql = "select " + SELECT_KEY + " from channel"
        val list = NPSDB.db.selectList(ChannelDto::class.java, sql)
        return list
    }

    /**
     * 更新一条数据
     */
    fun update(dto: ChannelDto) {
        val sql =
            "update channel set name = ?,type = ?,server_port=?,target_port=?,enable_state=?,security_state=?,acl_state=?,remark=?,update_date=${System.currentTimeMillis()} where id = ? and update_date=?"
        NPSDB.db.exec(
            sql,
            dto.name,
            dto.type,
            dto.serverPort,
            dto.targetPort,
            dto.enableState,
            dto.securityState,
            dto.aclState,
            dto.remark,
            dto.id,
            dto.updateDate
        )
    }

    /**
     * 同步入出网流量
     */
    fun setDataLen(dto: ChannelDto) {
        val sql = "update channel set in_data_total = ?,out_data_total=? where id = ?"
        NPSDB.db.exec(sql, dto.inDataTotal, dto.outDataTotal, dto.id)
    }

    /**
     * @TODO: 删除数据流量统计信息
     * 通过id删除一条数据
     * @param id 隧道id
     */
    fun delete(id: Int) {
        val sql = "delete from channel where id = ?"
        NPSDB.db.exec(sql, id)
    }

    /**
     * 删除某个客户端下所有的隧道
     * @param clientId 客户端ID
     */
    fun deleteByClient(clientId: Int) {
        val sql = "delete from channel where client_id = ?"
        NPSDB.db.exec(sql, clientId)
    }

    /**
     * 设置备注信息
     */
    fun setRemark(id: Int, remark: String) {
        val sql =
            "update channel set remark = ? where id = ?"
        NPSDB.db.exec(sql, remark, id)
    }

    /**
     * 获取所有隧道列表
     */
    fun search(dto: ChannelListSearchDto): List<ChannelSearchDto> {
        val sql = StringBuilder(
            "select " + SELECT_KEY +
                    ",client.name as clientName" +
                    " from channel left join client on channel.client_id = client.id where 1=1 "
        )

        if (dto.clientId != null) {
            sql.append(" and channel.client_id = ").append(dto.clientId)
        }

        if (dto.type != null) {
            sql.append(" and channel.type = ").append(dto.type)
        }
        sql.append(" order by id desc")
        return NPSDB.db.selectList(ChannelSearchDto::class.java, sql.toString())
    }

    /**
     * 获取所有激活的隧道列表
     */
    fun selectActiveByClientId(clientId: Int): List<ChannelDto> {
        val sql = "select" + SELECT_KEY +
                " from channel left join client on channel.client_id = client.id where channel.client_id = ? and client.enable_state = 1 and channel.enable_state = 1"
        return NPSDB.db.selectList(ChannelDto::class.java, sql, clientId)
    }

    /**
     * 获取客户端下所有的隧道id列表
     */
    fun selectIdByClientId(clientId: Int): List<Int> {
        val sql = "select id from channel where client_id = ?"
        return NPSDB.db.selectList(ChannelDto::class.java, sql, clientId).map {
            it.id!!
        }
    }
}

File: dao/ChannelDao/ChannelDao.go
// 隧道数据操作
package ChannelDao

				import (
					"DairoNPS/DebugTimer"
				)

import (
	"DairoNPS/dao/dto"
	"DairoNPS/util/DBUtil"
	"strconv"
)

func init() {
DebugTimer.Add56()
	ClearError()
}

/**
 * 添加一条隧道
 */
func Add(dto *dto.ChannelDto) {
DebugTimer.Add57()
	sql :=
		"insert into channel(clientId,name,mode,serverPort,targetPort,securityState,aclState,remark)values(?,?,?,?,?,?,?,?)"
	id := DBUtil.InsertIgnoreError(
		sql,
		dto.ClientId,
		dto.Name,
		dto.Mode,
		dto.ServerPort,
		dto.TargetPort,
		dto.SecurityState,
		dto.AclState,
		dto.Remark,
	)
	dto.Id = int(id)
}

/**
 * 通过id获取一条数据
 * @param id 隧道id
 * @return 隧道Dto
 */
func SelectOne(id int) *dto.ChannelDto {
DebugTimer.Add58()
	sql := "select * from channel where id = ?"
	return DBUtil.SelectOne[dto.ChannelDto](sql, id)
}

// 通过端口查询一条数据
func SelectByPort(port int) *dto.ChannelDto {
DebugTimer.Add59()
	sql := "select * from channel where serverPort = ?"
	return DBUtil.SelectOne[dto.ChannelDto](sql, port)
}

/**
 * 获取所有数据
 * @return 隧道Dto
 */
func SelectAll() []*dto.ChannelDto {
DebugTimer.Add60()
	sql := "select * from channel"
	return DBUtil.SelectList[dto.ChannelDto](sql)
}

/**
 * 更新一条数据
 */
func Update(dto *dto.ChannelDto) {
DebugTimer.Add61()
	sql :=
		"update channel set name = ?,mode = ?,serverPort=?,targetPort=?,securityState=?,aclState=?,remark=? where id = ?"
	DBUtil.ExecIgnoreError(
		sql,
		dto.Name,
		dto.Mode,
		dto.ServerPort,
		dto.TargetPort,
		dto.SecurityState,
		dto.AclState,
		dto.Remark,
		dto.Id,
	)
}

/**
 * 同步入出网流量
 */
func SetDataSize(id int, inData int64, outData int64) {
DebugTimer.Add62()
	sql := "update channel set inData = ?,outData=? where id = ?"
	DBUtil.ExecIgnoreError(sql, inData, outData, id)
}

/**
 * @TODO: 删除数据流量统计信息
 * 通过id删除一条数据
 * @param id 隧道id
 */
func Delete(id int) {
DebugTimer.Add63()
	sql := "delete from channel where id = ?"
	DBUtil.ExecIgnoreError(sql, id)
}

/**
 * 删除某个客户端下所有的隧道
 * @param clientId 客户端ID
 */
func DeleteByClient(clientId int) {
DebugTimer.Add64()
	sql := "delete from channel where clientId = ?"
	DBUtil.ExecIgnoreError(sql, clientId)
}

/**
 * 设置备注信息
 */
func SetRemark(id int, remark string) {
DebugTimer.Add65()
	sql := "update channel set remark = ? where id = ?"
	DBUtil.ExecIgnoreError(sql, remark, id)
}

/**
 * 获取所有隧道列表
 */
func Search(searchDto dto.ChannelListSearchDto) []*dto.ChannelSearchDto {
DebugTimer.Add66()
	sql := "select channel.*,client.name as clientName" +
		" from channel left join client on channel.client_id = client.id where 1=1 "

	if searchDto.ClientId != 0 {
DebugTimer.Add67()
		sql += " and channel.client_id = " + strconv.Itoa(searchDto.ClientId)
	}

	if searchDto.Mode != 0 {
DebugTimer.Add68()
		sql += " and channel.mode = " + strconv.Itoa(searchDto.Mode)
	}
	sql += " order by id desc"
	return DBUtil.SelectList[dto.ChannelSearchDto](sql)
}

/**
 * 获取所有激活的隧道列表
 */
func SelectActiveByClientId(clientId int) []*dto.ChannelDto {
DebugTimer.Add69()
	sql := "select channel.* from channel left join client on channel.clientId = client.id where channel.clientId = ? and client.enableState = 1 and channel.enableState = 1"
	return DBUtil.SelectList[dto.ChannelDto](sql, clientId)
}

/**
 * 获取客户端下所有的隧道id列表
 */
func SelectByClientId(clientId int) []*dto.ChannelDto {
DebugTimer.Add70()
	sql := "select * from channel where clientId = ?"
	return DBUtil.SelectList[dto.ChannelDto](sql, clientId)
}

/**
 * 获取客户端下所有的隧道id列表
 */
func SelectIdByClientId(clientId int) []int {
DebugTimer.Add71()
	sql := "select id from channel where clientId = ?"
	list := DBUtil.SelectList[dto.ChannelDto](sql, clientId)
	ids := make([]int, 0)
	for _, it := range list {
DebugTimer.Add72()
		ids = append(ids, it.Id)
	}
	return ids
}

// 设置可用状态
func SetEnableState(id int, state int) {
DebugTimer.Add73()
	sql := "update channel set enableState = ? where id = ?"
	DBUtil.ExecIgnoreError(sql, state, id)
}

// 设置错误信息
func SetError(id int, error *string) {
DebugTimer.Add74()
	sql := "update channel set error = ? where id = ?"
	DBUtil.ExecIgnoreError(sql, error, id)
}

// 清空错误信息
func ClearError() {
DebugTimer.Add75()
	sql := "update channel set error = null"
	DBUtil.ExecIgnoreError(sql)
}
